include/ginkgo/core/synthesizer/containers.hpp

```diff
@@ -46,24 +47,55 @@ namespace gko {
 namespace syn {
 
 
+/**
+ * value_list records several values with the same type in template.
+ *
+ * @tparam T  the value type of the list
+ * @tparam T...  the values in the list
+ */
 template <typename T, T... Values>
 struct value_list {};
 
 
+/**
+ * type_list records several types in template
+ *
+ * @tparam ...Types  the types in the list
+ */
 template <typename... Types>
 struct type_list {};
 
 
+/**
+ * range records start, end, step in template
+ *
+ * @tparam int  start of range
+ * @tparam int  end of range
+ * @tparam int  step of range. default is 1
+ */
 template <int Start, int End, int Step = 1>
 struct range {};
 
 
 namespace detail {
 
 
+/**
+ * concatenate_impl base type
+ *
+ * @tparam List1  the first List
+ * @tparam List2  the second List
+ */
 template <typename List1, typename List2>
 struct concatenate_impl;
 
+/**
+ * concatenate_impl specializes for two value_list with the same value type.
+ *
+ * @tparam T  the value type of two value_list
+ * @tparam T...  the values of the first list
+ * @tparam T...  the values of the second list
```

Comment:
nit: Wouldn't the `tparam` be the name of the types (and not just the types)?
```suggestion
 * @tparam Values1  the values of the first list
 * @tparam Values2  the values of the second list
```